File: design-patterns/golang/creational/factory_method/factory_method.go
/*!
* \brief. Factory method.
*
* \pattern. Define an interface for creating an object,
*         but let subclasses decide which class to
*         instantiate.Factory Method lets a class defer
*         instantiation to subclasses
*
*          Animal     < = >   AnimalFactory
*            |
*        AnimalBase            |        |
*         |     |
*       Lion   Tiger    LionFactory   TigerFactory
*
* \note. Inheritance does not exist in Go, so an anonymous combination is used.
 */

package main

import (
	"fmt"
)

// 方法接口
type Animal interface {
	SetAge(int)
	GetInfo()
}

// AnimalFactory, 工厂接口，Create返回的是Animal接口.
type AnimalFactory interface {
	Create() Animal
}

// AnimalBase. Animal接口实现的基类，封装公用方法
type AnimalBase struct {
	age int
}

func (o *AnimalBase) SetAge(age int) {
	o.age = age
}

// LionFactory.
type LionFactory struct{}

func (LionFactory) Create() Animal {
	return &Lion{
		AnimalBase: &AnimalBase{},
	}
}

// Lion. 相当于Animal接口下AnimalBase的派生类
type Lion struct {
	*AnimalBase
}

func (o Lion) GetInfo() {
	fmt.Println("Species: Lion, Age:", o.age)
}

// TigerFactory.
type TigerFactory struct{}

func (TigerFactory) Create() Animal {
	return &Tiger{
		AnimalBase: &AnimalBase{},
	}
}

// Tiger.
type Tiger struct {
	*AnimalBase
}

func (o Tiger) GetInfo() {
	fmt.Println("Species: Tiger, Age:", o.age)
}

//
func main() {
	var factory AnimalFactory

	factory = LionFactory{}
	animal := factory.Create()
	animal.SetAge(5)
	animal.GetInfo()

	factory = TigerFactory{}
	animal = factory.Create()
	animal.GetInfo()
}

File: design-patterns/python/creational/factory_method.py
#!/usr/bin/python
# -*- coding : utf-8 -*-

""" 
brief:
    Factory method. Use multiple factories for 
  multiple products

pattern:
  Factory method:
    Define an interface for creating an object,
  but let subclasses decide which class to
  instantiate.Factory Method lets a class defer
  instantiation to subclasses.
 
    AbstractProduct     < = >        AbstractFactory
      |         |                     |          |
  ProductA  ProductB    FactoryA(for ProductA)  FactoryB(for ProductB)
"""

# AbstractProduct
class Animal:
  def __init__(self):
    self.boundary = 'Animal'
    self.state = self.boundary
  def __str__(self):
    return self.state
	
# Product
class Lion(Animal):
  def __init__(self):
    super(Lion, self).__init__();
    self.state = self.boundary + ' - Lion'
  def __str__(self):
    return self.state
	
class Tiger(Animal):
  def __init__(self):
    super(Tiger, self).__init__();
    self.state = self.boundary + ' - Tiger'	
  def __str__(self):
    return self.state

# AbstractFactory
class Factory():
    def create_animal(self):
        pass
        
# Factory
class LionFactory(Factory):
  def __init__(self):
    print('Initialize LionFactory.')
  def create_animal(self):
      return Lion()
 
class TigerFactory(Factory):
  def __init__(self):
    print('Initialize TigerFactory.')
  def create_animal(self):
      return Tiger()
      
if __name__ == '__main__':
  lion_factory = LionFactory()
  tiger_factory = TigerFactory()
  
  obj1 = lion_factory.create_animal()
  obj2 = tiger_factory.create_animal()
	
  print('obj1: {0}'.format(obj1))
  print('obj2: {0}'.format(obj2))